File: src/claude_fix.py
"""Claude Code fix orchestrator for auto-review workflow."""

import json
import logging
import subprocess
from dataclasses import dataclass

from .gemini_review import ReviewResult, format_fix_prompt

logger = logging.getLogger(__name__)


@dataclass
class FixResult:
    success: bool
    changed: bool  # Were files actually modified?
    error: str = ""


def get_modified_files(work_dir: str) -> set[str]:
    """Get set of modified files in the working directory."""
    try:
        result = subprocess.run(
            ["git", "diff", "--name-only"],
            cwd=work_dir,
            capture_output=True,
            text=True,
            timeout=10,
        )
        return set(result.stdout.strip().split("\n")) if result.stdout.strip() else set()
    except Exception:
        return set()


def fix_issues(work_dir: str, review: ReviewResult, mr_iid: int) -> FixResult:
    """Run Claude Code to fix review issues, then commit and push."""
    prompt = format_fix_prompt(review)

    if "No critical or warning issues found" in prompt:
        return FixResult(success=True, changed=False)

    logger.info(f"Starting Claude Code fix for MR !{mr_iid}...")

    # Get current state
    before_changes = get_modified_files(work_dir)

    try:
        result = subprocess.run(
            [
                "claude",
                "-p",
                prompt,
                "--permission-mode",
                "acceptEdits",
                "--allowed-tools",
                "Read Edit Bash",
                "--output-format",
                "json",
                "--no-session-persistence",
            ],
            cwd=work_dir,
            capture_output=True,
            text=True,
            timeout=600,  # 10 minutes
        )

        if result.returncode != 0:
            stderr = result.stderr[:1000]
            logger.error(f"Claude Code exited with code {result.returncode}: {stderr}")
            return FixResult(success=False, changed=False, error=f"Claude Code failed: {stderr}")

        logger.info("Claude Code completed successfully.")

    except subprocess.TimeoutExpired:
        logger.error("Claude Code timed out after 10 minutes")
        return FixResult(success=False, changed=False, error="Claude Code timed out")

    # Check if files were actually modified
    after_changes = get_modified_files(work_dir)
    new_or_modified = after_changes - before_changes

    if not new_or_modified:
        logger.info("No files were modified by Claude Code.")
        return FixResult(success=True, changed=False)

    logger.info(f"Files modified: {', '.join(sorted(new_or_modified))}")

    # Commit and push
    try:
        subprocess.run(["git", "add", "-A"], cwd=work_dir, check=True, capture_output=True, timeout=30)
        subprocess.run(
            ["git", "commit", "-m", f"auto-fix: address review comments for MR !{mr_iid}"],
            cwd=work_dir,
            check=True,
            capture_output=True,
            text=True,
            timeout=30,
        )
        subprocess.run(["git", "push"], cwd=work_dir, check=True, capture_output=True, timeout=60)
        logger.info("Changes committed and pushed successfully.")
        return FixResult(success=True, changed=True)

    except subprocess.CalledProcessError as e:
        stderr = e.stderr[:500] if hasattr(e, "stderr") and e.stderr else "unknown error"
        logger.error(f"Git operation failed: {stderr}")
        return FixResult(success=False, changed=False, error=f"Git operation failed: {stderr}")

File: src/claude-fix.ts
import { exec } from 'node:child_process';
import { promisify } from 'node:util';
import { formatFixPrompt, type ReviewResult } from './gemini-review.js';

const execAsync = promisify(exec);

export interface FixResult {
  success: boolean;
  changed: boolean;
  error?: string;
}

async function getModifiedFiles(workDir: string): Promise<Set<string>> {
  try {
    const { stdout } = await execAsync('git diff --name-only', { cwd: workDir, timeout: 10_000 });
    const files = stdout.trim().split('\n').filter(Boolean);
    return new Set(files);
  } catch {
    return new Set();
  }
}

export async function fixIssues(
  workDir: string,
  review: ReviewResult,
  mrIid: number,
): Promise<FixResult> {
  const prompt = formatFixPrompt(review);
  if (prompt.includes('未发现需要自动修复')) {
    return { success: true, changed: false };
  }

  console.log(`开始 Claude Code 修复 MR !${mrIid}...`);

  const beforeChanges = await getModifiedFiles(workDir);

  try {
    await execAsync(
      `claude -p "${prompt.replace(/"/g, '\\"')}" --permission-mode acceptEdits --allowed-tools "Read Edit Bash" --output-format json --no-session-persistence`,
      { cwd: workDir, timeout: 600_000 },
    );
    console.log('Claude Code 修复完成。');
  } catch (e: unknown) {
    const error = e instanceof Error ? e.message : String(e);
    console.error(`Claude Code 执行失败: ${error}`);
    return { success: false, changed: false, error };
  }

  const afterChanges = await getModifiedFiles(workDir);
  const newOrModified = new Set([...afterChanges].filter((f) => !beforeChanges.has(f)));

  if (newOrModified.size === 0) {
    console.log('Claude Code 未修改任何文件。');
    return { success: true, changed: false };
  }

  console.log(`修改的文件: ${[...newOrModified].sort().join(', ')}`);

  try {
    await execAsync('git add -A', { cwd: workDir, timeout: 30_000 });
    await execAsync(`git commit -m "auto-fix: 修复 MR !${mrIid} 审查问题"`, { cwd: workDir, timeout: 30_000 });
    await execAsync('git push', { cwd: workDir, timeout: 60_000 });
    console.log('修复已提交并推送。');
    return { success: true, changed: true };
  } catch (e: unknown) {
    const error = e instanceof Error ? e.message : String(e);
    console.error(`Git 操作失败: ${error}`);
    return { success: false, changed: false, error: `Git 操作失败: ${error}` };
  }
}
